### PR TITLE
add test for a flow with latency according to #4201

### PR DESCRIPTION
**Pull request by andriidovhan**

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 1]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
         database.resetCosts()
     }
 
+    @Tidy
+    def "Able to create DEGRADED flow with LATENCY strategy if flowPath > max_latency"() {
+        given: "2 non-overlapping paths with 11 and 15 latency"
+        setLatencyForPaths(11, 15)
+
+        when: "Create a flow, maxLatency 5 and maxLatencyTier2 10"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:250]: description not related to context

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 2]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+        def flow = flowHelperV2.randomFlow(switchPair).tap {
+            allocateProtectedPath = false
+            maxLatency = 9
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:253]: can be 10?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 3]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+            maxLatencyTier2 = 16
+            pathComputationStrategy = PathComputationStrategy.LATENCY.toString()
+        }
+        northboundV2.addFlow(flow)
+
+        then: "Flow is created in DEGRADED state because flowPath doesn't satisfy max_latency value"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:259]: please add `but satisfies max_latency_tier2`, would make the intended behavior very clear

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 4]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+    @Tidy
+    def "Able to create DEGRADED flow with LATENCY strategy if flowPath > max_latency"() {
+        given: "2 non-overlapping paths with 11 and 15 latency"
+        setLatencyForPaths(11, 15)
+
+        when: "Create a flow, maxLatency 5 and maxLatencyTier2 10"
+        def flow = flowHelperV2.randomFlow(switchPair).tap {
+            allocateProtectedPath = false
+            maxLatency = 9
+            maxLatencyTier2 = 16
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:254]: can be 12? I don't follow the chosen boundary values. Can you explain?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 5]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+        flow && flowHelperV2.deleteFlow(flow.flowId)
+    }
+
+    @Tidy
+    @Tags([LOW_PRIORITY])
+    def "Able to create a flow with LATENCY strategy when max_latency > pathLatency > max_latency_tier2"() {
+        given: "2 non-overlapping paths with 11 and 15 latency"
+        setLatencyForPaths(11, 15)
+
+        when: "Create a flow, maxLatency 5 and maxLatencyTier2 10"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:279]: same

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 6]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+        def flow = flowHelperV2.randomFlow(switchPair).tap {
+            allocateProtectedPath = false
+            maxLatency = 19
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:282]: 12?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy [hunk 7]
```diff
@@ -242,6 +242,58 @@ class MaxLatencySpec extends HealthCheckSpecification {
...
+            maxLatencyTier2 = 8
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:283]: 10?

**Review by rtretyak — CHANGES_REQUESTED** [review]

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy (hunk 6 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy]: There was an issue when we were not able to create a flow in case `max_latency > max_latency_tier2`
This test was added just to cover this case.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy (hunk 4 above)
Comment (rtretyak) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:264]: I still don't get it. This should work the same with `maxLatencyTier2 = 12`. Second path with '15' latency is not even used in this test and we are solely focused on the '11' latency path. So if should not satisfy the `maxLatency=10` restriction but should satisfy the tier2 restriction. Is it the case being tested here? If yes, then `16` value completely obfuscates the goal of the test and messes with the second '15' path which is not used in the test

Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy:264]: Agree, it is not clear. I didn't think from this side.
The system always tries to find a path with the best latency for a flow with the `latency` strategy.
The `max_latency` and `max_latency_tier2` fields are used for setting the status of the flow.

for example:
We have two paths with 11 and 15 latency
also we have two flows (flow_1: mL=10, mlT2=12; flow_2: mL=14, mlT2=16)
Both flows are build via the path with 11 latency.
But the status of the flows are different: flow_1 - degraded, flow_2-up


The `maxLatencyTier2` value in the test was changed to 12, now it should clear.

Comment (rtretyak) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/MaxLatencySpec.groovy]: Good example. The code also looks clear now 👍

**Review by rtretyak — APPROVED** [review]
Maybe not related to this particular PR, but I'm now interested in how the system will behave in cases:
1. Latency value on the flow is the same as latency value on the path
2. ml = mlT2 = path

**Comment by andriidovhan** [comment]
> Maybe not related to this particular PR, but I'm now interested in how the system will behave in cases:
> 
>     1. Latency value on the flow is the same as latency value on the path
> 
>     2. ml = mlT2 = path

To tell the truth, I don't know.
I will check it.

**Comment by andriidovhan** [comment]
> > Maybe not related to this particular PR, but I'm now interested in how the system will behave in cases:
> > ```
> > 1. Latency value on the flow is the same as latency value on the path
> > 
> > 2. ml = mlT2 = path
> > ```
 In these cases, the flow will be built via a path that has the same latency as ml and the flow will be in UP state.
